File: modules/crypto-core/src/main/java/com/shterneregen/securelan/crypto/keystore/KeyStoreService.java
package com.shterneregen.securelan.crypto.keystore;

import java.io.IOException;
import java.nio.file.Path;
import java.security.GeneralSecurityException;
import java.security.KeyStore;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.cert.Certificate;

public interface KeyStoreService {
    KeyStore createPkcs12();

    KeyStore load(Path path, char[] password) throws IOException, GeneralSecurityException;

    void store(KeyStore keyStore, Path path, char[] password) throws IOException, GeneralSecurityException;

    void setPrivateKeyEntry(KeyStore keyStore,
                            String alias,
                            PrivateKey privateKey,
                            char[] entryPassword,
                            Certificate[] chain) throws GeneralSecurityException;

    void setCertificateEntry(KeyStore keyStore, String alias, Certificate certificate) throws GeneralSecurityException;

    PrivateKey getPrivateKey(KeyStore keyStore, String alias, char[] entryPassword) throws GeneralSecurityException;

    PublicKey getPublicKey(KeyStore keyStore, String alias) throws GeneralSecurityException;

    Certificate getCertificate(KeyStore keyStore, String alias) throws GeneralSecurityException;
}

File: modules/crypto-core/src/main/kotlin/com/shterneregen/securelan/crypto/keystore/KeyStoreService.kt
package com.shterneregen.securelan.crypto.keystore

import java.io.IOException
import java.nio.file.Path
import java.security.GeneralSecurityException
import java.security.KeyStore
import java.security.PrivateKey
import java.security.PublicKey
import java.security.cert.Certificate

interface KeyStoreService {
    fun createPkcs12(): KeyStore

    @Throws(IOException::class, GeneralSecurityException::class)
    fun load(path: Path, password: CharArray): KeyStore

    @Throws(IOException::class, GeneralSecurityException::class)
    fun store(keyStore: KeyStore, path: Path, password: CharArray)

    @Throws(GeneralSecurityException::class)
    fun setPrivateKeyEntry(
        keyStore: KeyStore,
        alias: String,
        privateKey: PrivateKey,
        entryPassword: CharArray,
        chain: Array<Certificate>,
    )

    @Throws(GeneralSecurityException::class)
    fun setCertificateEntry(keyStore: KeyStore, alias: String, certificate: Certificate)

    @Throws(GeneralSecurityException::class)
    fun getPrivateKey(keyStore: KeyStore, alias: String, entryPassword: CharArray): PrivateKey

    @Throws(GeneralSecurityException::class)
    fun getPublicKey(keyStore: KeyStore, alias: String): PublicKey?

    @Throws(GeneralSecurityException::class)
    fun getCertificate(keyStore: KeyStore, alias: String): Certificate?
}
